core/crypto/src/key_file.rs

```diff
@@ -22,10 +43,259 @@ impl KeyFile {
+
+    pub fn from_file(path: &Path) -> Self {  
         let mut file = File::open(path).expect("Could not open key file.");
         let mut content = String::new();
         file.read_to_string(&mut content).expect("Could not read from key file.");
         serde_json::from_str(&content).expect("Failed to deserialize KeyFile")
     }
+    
+    // decrypts a file and return an instance of the file
+    pub fn from_encrypted_file(path: &Path, key : &Key, key_file: KeyFile) -> Self {
+        let ssb = key_file.ssb;
+        let decrypted_secretkey = ssb.decrypt(&secretKeyToVec(key_file.secret_key), &key);
+        KeyFile {
+            account_id: key_file.account_id,
+            public_key: key_file.public_key,
+            secret_key: secretKeyFromVec(decrypted_secretkey),
+            key_hash: KeyHash::default(),
+            ssb: SodiumoxideSecretBox::default(),
+        }
+    }
+
+    pub fn is_encrypted(&self) -> bool{
+        !self.key_hash.eq(&KeyHash::default())
+    }
+
+    // if a private key is encrypted , ask for a password in console
+    // if an entered password matches a password entered in step 3, decode an encrypted password otherwise keep asking for a correct password
+    pub fn from_encrypted_file_with_password(path : &Path, input: &str) -> Self  {  
+        //reading file
+        let mut key_file = KeyFile::from_file(path);
+        //password is encrypted proceed further, if not Ok(key_file)
+        if key_file.is_encrypted() {
+        //ask for a password in console
+        //infinite loop
+            let password_input = PasswordInput::process_input(input).unwrap(); 
+            let password_slice = password_input.0.as_bytes();
+            // password is not empty decode a file with a passwordm if password empty do nothing
+            if !password_input.0.is_empty() { 
+                let decryption_keyhash = KeyHash::new(password_slice);
+                if decryption_keyhash.eq(&key_file.key_hash) { // if hashes of passwords match then decrypt Keyfile.secret_key
+                        let decryption_key = SodiumoxideSecretBox::gen_key(password_slice).unwrap();
+                        let cloned_keyfile = key_file.clone();
+                        key_file = KeyFile::from_encrypted_file(path, &decryption_key, cloned_keyfile);
+                    }
+                }
+            }
+        key_file
+    }
+}
+
+fn secretKeyToVec(secret_key: SecretKey) -> Vec<u8> {
+    match secret_key {
+        SecretKey::ED25519(sk) => { sk.0.to_vec() },
+        SecretKey::SECP256K1(sk) => unsafe {
+            mem::transmute_copy::<secp256k1::key::SecretKey, [u8; 32]>(&sk).to_vec()
+        },
+    }
+}
+
+fn secretKeyFromVec(vec: Vec<u8>) -> SecretKey {
+    if vec.len() == 64 {
+        SecretKey::ED25519(
+            unsafe {
+                mem::transmute_copy::<Vec<u8>, sodiumoxide::crypto::sign::ed25519::SecretKey>(&vec)
+            }
+        )
+    }
+    else {
+        let mut array = [0u8; 32];
```

Comment:
@nearmax [Salsa](http://www.crypto-it.net/eng/symmetric/salsa20.html?tab=1) stream cipher produces 64 byte long ciphertext. Although It works fine with SecretKey::ED25519 , but it fails with  SecretKey::SECP256K1  that is only 32 byte long.  To resolve this issue I suggest to use the following data structure pub struct SecretKeyView(Vec<u8>) instead of KeyFile.secret_key field. I would like to know your feedback regarding this matter.